barretenberg/cpp/src/barretenberg/vm/avm/trace/trace.cpp

```diff
@@ -1804,10 +1803,15 @@ AvmError AvmTraceBuilder::op_calldata_copy(uint8_t indirect,
     const uint32_t cd_offset = static_cast<uint32_t>(unconstrained_read_from_memory(cd_offset_resolved));
     const uint32_t copy_size = static_cast<uint32_t>(unconstrained_read_from_memory(copy_size_offset_resolved));
 
+    auto calldata = current_ext_call_ctx.calldata;
     if (is_ok(error)) {
-        slice_trace_builder.create_calldata_copy_slice(
-            calldata, clk, call_ptr, cd_offset, copy_size, dst_offset_resolved);
-        mem_trace_builder.write_calldata_copy(calldata, clk, call_ptr, cd_offset, copy_size, dst_offset_resolved);
+        if (current_ext_call_ctx.context_id == 0) {
+            slice_trace_builder.create_calldata_copy_slice(
+                calldata, clk, call_ptr, cd_offset, copy_size, dst_offset_resolved);
+            mem_trace_builder.write_calldata_copy(calldata, clk, call_ptr, cd_offset, copy_size, dst_offset_resolved);
+        } else {
+            write_slice_to_memory(dst_offset_resolved, AvmMemoryTag::FF, calldata);
+        }
     }
 
```

Comment:
context_id == 0 is essentially the top level context for the enqueued call - so this is the logic that writes to the calldata column, otherwise we just write to mainline memory